examples/example_vm/example_vm.cpp

```diff
@@ -0,0 +1,300 @@
+            break;
+        }
+
+        case OP_ADDRESS:
+        {
+            evmc_address address = msg->destination;
+            evmc_uint256be value = {};
+            std::memcpy(&value.bytes[12], address.bytes, sizeof(address.bytes));
+            stack.push(value);
+            break;
+        }
+
+        case OP_CALLDATALOAD:
+        {
+            size_t offset = stack.pop().bytes[31];
+            evmc_uint256be value = {};
+
+            size_t copy_size = 0;
+            if (offset < msg->input_size)
+                copy_size = msg->input_size - offset;  // Partial copy.
+            if (copy_size > sizeof(value))
+                copy_size = sizeof(value);  // Full copy.
+            std::memcpy(value.bytes, &msg->input_data[offset], copy_size);
+
+            stack.push(value);
+            break;
+        }
+
+        case OP_NUMBER:
+        {
+            evmc_uint256be value = {};
+            value.bytes[31] = static_cast<uint8_t>(host->get_tx_context(context).block_number);
+            stack.push(value);
+            break;
+        }
+
+        case OP_MSTORE:
+        {
+            uint8_t index = stack.pop().bytes[31];
+            evmc_uint256be value = stack.pop();
+            memory.set(index, value.bytes, sizeof(value));
+            break;
+        }
+
+        case OP_SLOAD:
+        {
+            evmc_uint256be index = stack.pop();
+            evmc_uint256be value = host->get_storage(context, &msg->destination, &index);
+            stack.push(value);
+            break;
+        }
+
+        case OP_SSTORE:
+        {
+            evmc_uint256be index = stack.pop();
+            evmc_uint256be value = stack.pop();
+            host->set_storage(context, &msg->destination, &index, &value);
+            break;
+        }
+
+        case OP_MSIZE:
+        {
+            uint8_t size_as_byte = static_cast<uint8_t>(memory.size);
+            evmc_uint256be value = {};
+            value.bytes[31] = size_as_byte;
+            stack.push(value);
+            break;
+        }
+
+        case OP_PUSH1:
+        {
+            uint8_t byte = code[pc + 1];
+            pc++;
+            evmc_uint256be value = {};
+            value.bytes[31] = byte;
+            stack.push(value);
+            break;
+        }
+
+        case OP_DUP1:
+        {
+            evmc_uint256be value = stack.pop();
+            stack.push(value);
+            stack.push(value);
+            break;
+        }
+
+        case OP_CALL:
+        {
+            evmc_message call_msg = {};
+            call_msg.gas = stack.pop().bytes[31];
+            evmc_uint256be a = stack.pop();
+            std::memcpy(call_msg.destination.bytes, &a.bytes[12], sizeof(call_msg.destination));
+            call_msg.value = stack.pop();
+            uint8_t call_input_offset = stack.pop().bytes[31];
+            call_msg.input_data = &memory.data[call_input_offset];
+            call_msg.input_size = stack.pop().bytes[31];
```

Comment:
This does not updates `msize`.

Reply:
No it does not. Memory reads never update the msize.

Reply:
Are you sure? I think every memory touch extends the memory to that point. Just check the implementation of `op_return` in evmone as an example.

Reply:
I mean this example ignores it. EVM does not.

Reply:
Okay my point was perhaps do it in this example too? It might be just confusing to have an partial-partial-EVM implementation e.g. partially implementing some portion of the opcodes.